lib/storage/metricnamestats/tracker.go

```diff
@@ -0,0 +1,572 @@
+	var result StatsResult
+	if mt == nil {
+		return result
+	}
+	mt.mu.RLock()
+
+	result = mt.getStatsLocked(limit, func(sk *statKey, si *statItem) bool {
+		if sk.accountID != accountID || sk.projectID != projectID {
+			return false
+		}
+		if le >= 0 && int(si.requestsCount.Load()) > le {
+			return false
+		}
+		if len(matchPattern) > 0 && !strings.Contains(sk.metricName, matchPattern) {
+			return false
+		}
+		return true
+	})
+	mt.mu.RUnlock()
+
+	result.sort()
+	return result
+}
+
+// GetStats returns stats response for the tracked metrics
+//
+// DeduplicateMergeRecords must be called at cluster version on returned result.
+func (mt *Tracker) GetStats(limit, le int, matchPattern string) StatsResult {
+	var result StatsResult
+	if mt == nil {
+		return result
+	}
+	mt.mu.RLock()
+	defer mt.mu.RUnlock()
+
+	result = mt.getStatsLocked(limit, func(sk *statKey, si *statItem) bool {
+		if le >= 0 && int(si.requestsCount.Load()) > le {
+			return false
+		}
+		if len(matchPattern) > 0 && !strings.Contains(sk.metricName, matchPattern) {
+			return false
+		}
+		return true
+	})
+
+	result.sort()
+	return result
+}
+
+func (mt *Tracker) getStatsLocked(limit int, predicate func(sk *statKey, si *statItem) bool) StatsResult {
+	var result StatsResult
+
+	result.CollectedSinceTs = mt.creationTs.Load()
+	result.TotalRecords = mt.currentItemsCount.Load()
+	result.MaxSizeBytes = mt.maxSizeBytes
+	result.CurrentSizeBytes = mt.currentSizeBytes.Load()
+
+	for sk, si := range mt.store {
+		if len(result.Records) >= limit {
+			return result
+		}
+		if predicate(&sk, si) {
+			result.Records = append(result.Records, StatRecord{
+				MetricName:    sk.metricName,
+				RequestsCount: si.requestsCount.Load(),
+				LastRequestTs: si.lastRequestTs.Load(),
+			})
+		}
+	}
+	return result
+}
+
+// StatsResult defines stats result for GetStats request
+type StatsResult struct {
+	CollectedSinceTs uint64
+	TotalRecords     uint64
+	MaxSizeBytes     uint64
+	CurrentSizeBytes uint64
+	Records          []StatRecord
+}
+
+// StatRecord defines stat record for given metric name
+type StatRecord struct {
+	MetricName    string
+	RequestsCount uint64
+	LastRequestTs uint64
+}
+
+func (sr *StatsResult) sort() {
+	sort.Slice(sr.Records, func(i, j int) bool {
+		return sr.Records[i].MetricName < sr.Records[j].MetricName
+	})
+}
+
+// DeduplicateMergeRecords performs merging duplicate records by metric name
+//
+// It is usual case for global tenant request at cluster version.
+func (sr *StatsResult) DeduplicateMergeRecords() {
+	if len(sr.Records) < 2 {
+		return
+	}
+	tmp := sr.Records[:0]
+	// deduplication uses sliding indexes
+	//
+	// records:
+	// [ 0    1    2    3    4    5    6   ]
+	//
+	// [ mn1, mn2, mn2, mn2, mn3, mn4, mn4 ]
+	//
+	//	0     1
+	//	0          2
+	//	           2    3
+	//	           2         4
+	//	           2              5
+	//	                          5    6
+	//
+	// result:
+	//
+	//	[0,1,4,5]
```

Comment:
shouldn't they have summarized values here?

Reply:
It's indexes of original array. I'm still not sure is visialization useful or not.